### PR TITLE
Add `generatedCounter` implicitly to claims

### PR DESCRIPTION
**Pull request by Baltoli**
This PR adds an additional pass to the spec-module compilation pipeline; this pass adds a claimed `_ => ?_` rewrite to the `generatedCounter` cell if the claim does not already mention that cell. Doing so fixes the issue identified in #3450, where the prover fails to assert that this cell _does not change_ in a claim containing fresh variables.

The code triggering the original bug has been added as a test case, as well as a derived test that explicitly states the effect of the rule on the counter cell.

There is one place that this PR requires a test change; in the PL tests there is a spec that relies on a circularity implicitly containing `<generatedCounter> _ </generatedCounter>` rather than `_ => ?_`. This can be addressed by either removing the explicit "counter goes up by precisely 2" part of the spec, or by asserting that the counter doesn't change in the circularity. This PR chooses the latter, but either will work.

Fixes https://github.com/runtimeverification/k/issues/3450

### CODE REVIEW

File: kernel/src/main/java/org/kframework/compile/AddImplicitComputationCell.java [hunk 1]
```diff
@@ -51,10 +50,10 @@ public Sentence apply(Module m, Sentence s) {
 
         if (s instanceof RuleOrClaim) {
             RuleOrClaim rule = (RuleOrClaim) s;
-            return rule.newInstance(apply(rule.body(), m), rule.requires(), rule.ensures(), rule.att());
+            return rule.newInstance(apply(rule.body(), m, rule instanceof Claim), rule.requires(), rule.ensures(), rule.att());
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/compile/AddImplicitComputationCell.java:53]: Nitpicking here.
If we are to differentiate between Rule and Claim, why not do it at the top level?
Have two different `apply` functions instead.

File: kernel/src/main/java/org/kframework/compile/AddImplicitCounterCell.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    private K apply(K term, Module m) {
+        List<K> items = IncompleteCellUtils.flattenCells(term);
+        if(alreadyHasGeneratedCounter(items)) {
+            return term;
+        }
+
+        KRewrite rew = KRewrite(
+                KApply(KLabel("#SemanticCastToInt"), KVariable("_")),
+                KApply(KLabel("#SemanticCastToInt"), KVariable("?_")));
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/compile/AddImplicitCounterCell.java]: We already have constants for these variables in `ResolveAnonVar.FRESH_ANON_VAR`.

File: kernel/src/main/java/org/kframework/compile/AddImplicitComputationCell.java (hunk 1 above)
Comment (Baltoli) [reply kernel/src/main/java/org/kframework/compile/AddImplicitComputationCell.java:53]: I think we'd end up duplicating more code in this call stack if we did the differentiation at the top level; not opposed in principle to doing so but the current phrasing was the cleanest I came up with.

The issue is really that the initial sentence is no longer fully needed when we `apply` as we're on;y looking at its body, so you'd need an `applyRule` and `applyClaim` method that only differ in how they call `shouldConsider`.

File: kernel/src/main/java/org/kframework/compile/IncompleteCellUtils.java [hunk 3]
```diff
@@ -65,7 +65,7 @@ public static List<K> getChildren(KApply cell) {
     private static KApply makeDots(boolean isOpen) {
         return isOpen ? dots : noDots;
     }
-    private static K makeBody(List<K> children) {
+    public static K makeBody(List<K> children) {
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/compile/IncompleteCellUtils.java]: I would make this package-private.

**Review by dwightguth — APPROVED** [review]
Looks good but one minor comment.

**Review by radumereuta — APPROVED** [review]

**Pull request merged**